assimilation_code/modules/utilities/utilities_mod.f90

```diff
@@ -480,6 +481,111 @@ subroutine check_namelist_read(iunit, iostat_in, nml_name)
+
+character(len=*), intent(in) :: input                 ! value from namelist, eg
+integer,          intent(in) :: integer_options(:)    ! possible integer values
+character(len=*), intent(in) :: string_options(:)     ! matching string values
+character(len=*), intent(in), optional :: context 
+integer                      :: flex_parser
+
+character(len=len_trim(input)) :: uppercase
+character(len=len(string_options)) :: possibility
+integer :: ios, iopt, candidate
+
+if ( .not. module_initialized ) call initialize_utilities
+
+flex_parser = MISSING_I
+
+! Try to read the input as an ASCII-coded integer ( i.e. '3') 
+
+read(input,*,iostat=ios) candidate
+if (ios == 0) then
+
+   flex_parser = candidate
+
+   ! Check for valid value
+   if (any(integer_options == flex_parser)) then
+      return
+   else
+      call flex_parser_error(input, integer_options, string_options, context)
+   endif
+endif
+
+! numeric conversion not possible, try to interpret as a string
+
+if (size(integer_options) /= size(string_options)) then
+   write(msgstring2,*)'size(integer_options) =',size(integer_options), &
+                  ' /= size( string_options) =',size( string_options)
+   call error_handler(E_ERR,'flex_parser',msgstring2,source,text2=context)
+endif
+
+uppercase = trim(input)
+call to_upper(uppercase)
+
+LOOP : do iopt = 1,size(integer_options)
+
+   possibility = string_options(iopt)
+   call to_upper(possibility)
+
+   if (uppercase == possibility) then
+      flex_parser = integer_options(iopt)
+      exit LOOP
+   endif
+
+enddo LOOP
+
+if (flex_parser == MISSING_I) &
+   call flex_parser_error(input, integer_options, string_options, context)
+
+end function flex_parser
+
+
+!-----------------------------------------------------------------------
+!> report any errors from the flex_parser routine
+
+subroutine flex_parser_error(input, integer_options, string_options, whofrom)
+
+character(len=*),           intent(in) :: input
+integer,                    intent(in) :: integer_options(:)
+character(len=*),           intent(in) :: string_options(:)
+character(len=*), optional, intent(in) :: whofrom
+
+character(len=256) :: string1
+integer :: iopt
+
+if (present(whofrom)) then
+   msgstring1 = trim(whofrom)//' no valid option found.' 
+else
+   msgstring1 = 'No valid option found.' 
+endif
+write(msgstring2,*)'input is "'//trim(input)//'"'
+write(msgstring3,*)'valid values are the following integers or matching character strings:'
+
+call error_handler(E_MSG, 'flex_parser', msgstring1, source, &
+           text2=msgstring2, text3=msgstring3)
+
+do iopt = 1,size(integer_options)
+   write(string1,*) integer_options(iopt), ' = "'//trim(string_options(iopt))//'"' 
+   call error_handler(E_MSG,'flex_parser',string1)
+enddo
+
+! Repeat the leading message (as an E_ERR) to help delineate the problem.
+call error_handler(E_ERR, 'flex_parser', msgstring1, source, &
+           text2=msgstring2, text3=msgstring3)
```

Comment:
this should only print msgstring1.  msgstring2 and 3 don't fit here.

Reply:
I disagree. I chose to repeat the message string since the intervening messages can be quite extensive and I wanted to make sure they recognize it is all part of the same problem.

Reply:
tim, please read the messages it's going to be printing?  strings 2 and 3 say "the following list" of items is valid.  the list precedes this message so it will sound like there's missing information.  repeating string 1 makes perfect sense to end on.

Reply:
You're right ... I can delete msgstring2, msgstring3 from that message.